cores/arduino/stm32/OpenAMP/mbox_ipcc.c

```diff
@@ -0,0 +1,203 @@
+ */
+
+/* Includes ------------------------------------------------------------------*/
+#include "virtio_config.h"
+#include <openamp/open_amp.h>
+#include "stm32_def.h"
+#include "openamp_conf.h"
+#include "mbox_ipcc.h"
+#include "core_debug.h"
+
+/* Private define ------------------------------------------------------------*/
+#define MASTER_CPU_ID       0
+#define REMOTE_CPU_ID       1
+#define IPCC_CPU_A7         MASTER_CPU_ID
+#define IPCC_CPU_M4         REMOTE_CPU_ID
+
+typedef enum {
+  RX_NO_MSG = 0,
+  RX_NEW_MSG = 1,
+  RX_BUF_FREE = 2
+} rx_status_t;
+
+/* Private variables ---------------------------------------------------------*/
+IPCC_HandleTypeDef hipcc;
+rx_status_t msg_received_ch1 = RX_NO_MSG;
+rx_status_t msg_received_ch2 = RX_NO_MSG;
+
+/* Private function prototypes -----------------------------------------------*/
+void IPCC_channel1_callback(IPCC_HandleTypeDef *hipcc, uint32_t ChannelIndex, IPCC_CHANNELDirTypeDef ChannelDir);
+void IPCC_channel2_callback(IPCC_HandleTypeDef *hipcc, uint32_t ChannelIndex, IPCC_CHANNELDirTypeDef ChannelDir);
+
+/**
+  * @brief  Initialize MAILBOX with IPCC peripheral
+  * @param  None
+  * @retval : Operation result
+  */
+int MAILBOX_Init(void)
+{
+  __HAL_RCC_IPCC_CLK_ENABLE();
+  HAL_NVIC_SetPriority(IPCC_RX1_IRQn, IPCC_IRQ_PRIO, IPCC_IRQ_SUBPRIO);
+  HAL_NVIC_EnableIRQ(IPCC_RX1_IRQn);
+
+  hipcc.Instance = IPCC;
+  if (HAL_IPCC_Init(&hipcc) != HAL_OK) {
+    Error_Handler();
+  }
+
+  if (HAL_IPCC_ActivateNotification(&hipcc, IPCC_CHANNEL_1, IPCC_CHANNEL_DIR_RX,
+                                    IPCC_channel1_callback) != HAL_OK) {
+    Error_Handler();
+    return -1;
+  }
+
+  if (HAL_IPCC_ActivateNotification(&hipcc, IPCC_CHANNEL_2, IPCC_CHANNEL_DIR_RX,
+                                    IPCC_channel2_callback) != HAL_OK) {
+    Error_Handler();
+    return -1;
+  }
+
+  return 0;
+}
+
+/**
+ * @brief  Process vring messages received from IPCC ISR
+ * @param vdev: virtio device
+ * @param vring_id: Vring ID
+ * @retval : Operation result
+ */
+int MAILBOX_Poll(struct virtio_device *vdev, uint32_t vring_id)
+{
+  int ret = -1;
+
+  switch (vring_id) {
+    case VRING0_ID:
+      if (msg_received_ch1 == RX_BUF_FREE) {
+        core_debug("Running vring0 (ch_1 buf free)\n");
+        /* This calls rpmsg_virtio_tx_callback(), which actually does nothing. */
+        rproc_virtio_notified(vdev, VRING0_ID);
+        msg_received_ch1 = RX_NO_MSG;
+        ret = 0;
+      }
+      break;
+    case VRING1_ID:
+      if (msg_received_ch2 == RX_NEW_MSG) {
+        core_debug("Running vring1 (ch_2 new msg)\n");
+        /**
+         * This calls rpmsg_virtio_rx_callback(), which calls virt_uart rx callback
+         * RING_NUM_BUFFS times at maximum.
+         */
+        rproc_virtio_notified(vdev, VRING1_ID);
+        msg_received_ch2 = RX_NO_MSG;
+
+        /* The OpenAMP framework does not notify for free buf: do it here */
+        rproc_virtio_notified(NULL, VRING1_ID);
```

Comment:
I copied this line from the CubeMP1 examples but I suspect it should be removed. I'm waiting for a response from a ST employee to confirm if it is okay to remove or it should be replaced by anything else: https://community.st.com/s/question/0D53W0000009CfFSAU/a-question-regarding-openamp-rpmsg-example

Reply:
OK. I guess @arnopo could answer to that 😉 
@arnopo any update on this, please?
Thanks in advance.

Reply:
Hi,
i confirm yes you can remove it.

Reply:
Thanks @arnopo
@kbumsik could you proceed to this update then we will able to merge.

Reply:
@fpistm Sure :)
@arnopo Thanks for confirming it. By the way, could I ask you any update regarding my report above?